### DISCUSSION TITLE
signed chars are not converted to str: is this intentional?

### DISCUSSION THREAD

**Discussion by YangJiao1996** [pybind/pybind11#4235] [Q&A] [answered]
I've been using `signed char` instead of undecorated `char` in my code, and in the generated bindings they seem to be converted to Python `int`s instead of `str`s.

Sample code would be something like this in C++:
```cpp
m.def("pass_signed_char", [](signed char c) { return c; });
```

And in Python:

```py
example.pass_char("h") # Will get 105 instead of "h"
```

I read through the documentation and the tests in `test_builtin_casters.cpp` and `test_builtin_casters.py` and neither of them are testing `signed char`s and whether the `char` character types will be converted to `str` instead of `int`. 

I wonder if this is by design or just an oversight?

**Comment by Skylion007** [comment, marked as answer]
We never designed any of the casters with the intent of using signed chars. You can always do a conversion of the return value to unsigned chars with a lambda or function wrapper. The fact that unsigned chars gets cast to anything at all is surprising, it seems they are implicitly convertible to int which is why they work at all: https://en.cppreference.com/w/cpp/language/implicit_conversion

**Reply by YangJiao1996** [reply]
This makes sense, as the [documentation](https://pybind11.readthedocs.io/en/stable/advanced/cast/overview.html) never mentions signed char or unsigned char.
